File: plugins/OrchidPosts/src/main/java/com/eden/orchid/posts/AuthorOptionExtractor.java
package com.eden.orchid.posts;

import com.eden.common.util.EdenPair;
import com.eden.orchid.api.OrchidContext;
import com.eden.orchid.api.converters.StringConverter;
import com.eden.orchid.api.options.OptionExtractor;
import com.google.inject.Provider;
import org.json.JSONObject;

import javax.inject.Inject;
import java.lang.reflect.Field;
import java.util.List;

/**
 * ### Destination Types
 *
 * | Field Type   | Annotation     | Default Value              |
 * |--------------|----------------|----------------------------|
 * | String       | @StringDefault | Annotation value() or null |
 * | String[]     | none           | Empty String[]             |
 */
public final class AuthorOptionExtractor extends OptionExtractor<Author> {

    private final Provider<OrchidContext> contextProvider;
    private final StringConverter converter;
    private final PostsModel postsModel;

    @Inject
    public AuthorOptionExtractor(Provider<OrchidContext> contextProvider, StringConverter converter, PostsModel postsModel) {
        super(1000);
        this.contextProvider = contextProvider;
        this.converter = converter;
        this.postsModel = postsModel;
    }

    @Override
    public boolean acceptsClass(Class clazz) {
        return clazz.equals(Author.class);
    }

    public String getValue(Object object) {
        return converter.convert(object).second;
    }

    @Override
    public Author getOption(Field field, JSONObject options, String key) {
        if(options.has(key)) {

            if(options.get(key) instanceof JSONObject) {
                Author author = new Author();
                author.extractOptions(contextProvider.get(), options.getJSONObject(key));
                return author;
            }
            else {
                EdenPair<Boolean, String> value = converter.convert(options.get(key));
                if(value.first) {
                    return postsModel.getAuthorByName(value.second);
                }
            }
        }

        return getDefaultValue(field);
    }

    @Override
    public Author getDefaultValue(Field field) {
        return null;
    }

    @Override
    public List<Author> getList(Field field, JSONObject options, String key) {
        return null;
    }

    @Override
    public Object[] getArray(Field field, JSONObject options, String key) {
        return null;
    }
}

File: plugins/OrchidPosts/src/main/kotlin/com/eden/orchid/posts/utils/AuthorOptionExtractor.kt
package com.eden.orchid.posts.utils

import com.eden.orchid.api.OrchidContext
import com.eden.orchid.api.converters.StringConverter
import com.eden.orchid.api.options.OptionExtractor
import com.eden.orchid.posts.model.Author
import com.eden.orchid.posts.model.PostsModel
import com.eden.orchid.utilities.resolve
import com.google.inject.Provider
import org.json.JSONObject
import java.lang.reflect.Field
import javax.inject.Inject

/**
 * ### Destination Types
 *
 * | Field Type   | Annotation     | Default Value              |
 * |--------------|----------------|----------------------------|
 * | String       | @StringDefault | Annotation value() or null |
 * | String[]     | none           | Empty String[]             |
 */
class AuthorOptionExtractor
@Inject
constructor(
        private val contextProvider: Provider<OrchidContext>,
        private val converter: StringConverter
) : OptionExtractor<Author>(1000) {

    override fun acceptsClass(clazz: Class<*>): Boolean {
        return clazz == Author::class.java
    }

    fun getValue(value: Any): String {
        return converter.convert(String::class.java, value).second
    }

    override fun getOption(field: Field, sourceObject: Any, key: String): Author? {
        val postsModel: PostsModel = contextProvider.get().resolve()

        if (sourceObject is JSONObject) {
            val author = Author()
            author.extractOptions(contextProvider.get(), sourceObject.toMap())
            return author
        }
        else {
            val value = converter.convert(String::class.java, sourceObject)
            if (value.first) {
                return postsModel.getAuthorByName(value.second)
            }
        }

        return null
    }

    override fun getDefaultValue(field: Field): Author? {
        return null
    }

}

